### ISSUE TITLE
Allow Negative Fees

### ISSUE THREAD

**Issue by aszepieniec** [Neptune-Crypto/neptune-core#251]
Currently, there is no mechanism for a transaction-upgrades-as-a-service provider to take payment for upgrading an external node's `ProofCollection`-transaction to a `SingleProof`-transaction. Allowing negative fees fixes this.

Specifically, the proposed solution involves the following points:
 - Drop the non-negativity and overflow checks on the fee in `NativeCurrency`.
 - Require that the output of a `Merge` has a non-negative fee.
 - Require that the block transaction has a non-negative fee.

The proposed solution addresses the payment issue as follows:
 - External node broadcasts a `ProofCollection` transaction with a certain transaction fee.
 - If the fee is large enough, the transaction-upgrader upgrades and merges it with a transaction of his own design, which allocates a portion of the fee to himself.
 - If the fee is not large enough, the transaction-upgrader ignores it. If the fee is too small, all transaction-upgraders on the network might decide to ignore it. That's fine: the initiator should broadcast a new transaction and bump the fee.

**Comment by dan-da** [comment]
Has this been implemented?   Can it be closed?

-------

anyway I will comment that it is an interesting solution and in particular I am curious about the scaling properties.  I feel it bears some discussion/scrutiny.

First, let's consider that generating a SingleProof ideally *should* be performed offline by the sender initiating the transaction, and thus should have zero impact on running nodes and their ability to process (verify) tx, nor on the max tx-per-sec of the entire network.

So then we consider that many/most senders will not be able to generate the proof themselves with present-day hardware, and must outsource that to more capable hardware "somehow".   Given this, it should seemingly remain a design goal of the outsourcing-mechanism that the proof-generation does not impact running nodes ability to verify tx, nor affect the max tx-per-sec of entire network.

I have doubts that this design achieves those goals.

In this design a ProofCollection gets added to the Mempool and any node with "free time" may decide to start processing it.

Let's consider an initial state with 100 nodes running and an empty mempool.   A tx (tx1) enters the mempool that needs upgrading and offers a fee.   So all 100 nodes begin processing the tx.   This is already a phenomenal duplication of resources, as only 1 device ever needs to generate this proof.   Ok, so while all 100 nodes are working on a proof for tx2, a second tx (tx2) enters the mempool, and a third (tx3).  But nobody processes either because all 100 are busy.  Finally some nodes finish with tx1 and begin processing tx3 because it has a higher fee than tx2.  At this moment we have a lot of nodes till working on tx1, some nodes starting on tx3, duplicating effort, and nobody processing tx2.

Clearly this approach duplicates a *lot* of work and is therefore going to process fewer tx-per-sec than an approach where each tx is processed only by a single node.

I think the primary benefit/appeal of this approach is that it does not require any coordination between (a) sender and upgrader nodes or (b) between upgrader nodes.   But it appears to me that this benefit will carry the cost of a drastically reduced network-wide capacity for processing tx.

As such, I would tend to favor/suggest a design where either:
1. sender coordinates/negotiates with upgrader node(s) and picks a single node.
2. upgrader nodes coordinate with eachother to avoid duplication of work.
3. upgrading is left as the sender's responsibility and not part of neptune-core.   A network/market could still be developed, but ti would be separate from neptune-core network.

(2) is probably easiest to add/adapt for current design.  simplest: announce to peers when starting an upgrade, don't begin if another node has already announced.

(3) is probably cleanest and most scalable long term as it places 0 burden on neptune-core nodes.

**Comment by Sword-Smith** [comment]
I agree with some of your objections (danger of duplicated work) but they can be solved through application logic and assuming rational actors. Whereas 17c4a8416fd5f8f31b1cf5b23cddc8e3fb218109 (and related commits) solve the problem of outsourcing proof-work from a consensus perspective.

**Comment by dan-da** [comment]
> they can be solved through application logic

ok, I will open an issue for this.

> Whereas https://github.com/Neptune-Crypto/neptune-core/commit/17c4a8416fd5f8f31b1cf5b23cddc8e3fb218109 (and related commits) solve the problem of outsourcing proof-work from a consensus perspective.

that means this can be closed, yes?

**Comment by Sword-Smith** [comment]
Closed by 17c4a8416fd5f8f31b1cf5b23cddc8e3fb218109, 9eb941f305f313f2abcd5f602ccdb964ed1ba893, and commits adding related tests: 54ffdea5050ca0c4746b868ce473111394303f22, 7f41efe981367ceb1f6e36f8346d480841282185, 0f74c71b740bf83a86e828404eaebf40140b3bda, c48acb70dd4fedfc56f587733a9ac1babb0e1994 etc.
